mapillary_tools/download_blurred.py

```diff
@@ -0,0 +1,400 @@
+
+        err_file_obj = open(
+            err_file, 'r') if err_file_exists is True else open(err_file, 'r+')
+
+        done_keys = done_file_obj.read().split('\n')
+        err_keys = err_file_obj.read().split('\n')
+        set_all = set(all_keys)
+        set_done = set(done_keys)
+        set_err = set(err_keys)
+
+        diff = list(set_all.difference(set_done).difference(set_err))
+        print("Download {} continued: {}/{}".format(download_id,
+                                                    len(diff), len(all_keys)))
+
+        image_keys = diff
+        done_file_obj.close()
+        err_file_obj.close()
+    else:
+        headers = get_headers_for_username(user_name)
+        image_keys = query_search_api(
+            headers,
+            organization_keys=organization_keys,
+            start_time=start_time,
+            end_time=end_time,
+            per_page='1000'
+        )
+
+        all_file_obj.writelines('\n'.join(image_keys))
+        all_file_obj.close()
+
+    done_file_obj = open(
+        done_file, 'w') if done_file_exists is True else open(done_file, 'a+')
+    err_file_obj = open(
+        err_file, 'w') if err_file_exists is True else open(err_file, 'w+')
+
+    # download begins here
+
+    if len(image_keys) == 0:
+        print('All images downloaded for this query. Remove the {}/all/err files to re-download.'.format(download_id))
+        return
+
+    lock = threading.Lock()
+
+    counter = {
+        'todo': 0,
+        'done': 0,
+        'err': 0,
+        'ok': 0
+    }
+
+    # Set up a queue for the downloads so threads can take downloads from
+    # it asynchronously
+    q = Queue(len(image_keys))
+
+    for i in image_keys:
+        q.put(i)
+
+    threads = []
+    monitor = None
+    try:
+        for i in range(number_threads):
+            t = BlurredOriginalsDownloader(
+                lock,
+                q,
+                output_folder,
+                headers,
+                counter,
+                done_file_obj,
+                err_file_obj)
+            threads.append(t)
+            t.start()
+
+        monitor = WorkerMonitor(threads, q)
+        monitor.start()
+
+        while True:
+            any_alive = False or monitor.is_alive()
+
+            for t in threads:
+                any_alive = (any_alive or t.is_alive())
+
+            if not any_alive:
+                break
+            time.sleep(0.5)
+
+    except (ServiceExit, KeyboardInterrupt):
+        monitor.shutdown_flag.set()
+
+        for t in threads:
+            t.shutdown_flag.set()
+
+        for t in threads:
+            t.join()
+            print('INFO: Thread stopped {}'.format(t))
+
+        monitor.join()
+        print('INFO: Thread stopped {}'.format(monitor))
+
+        all_file_obj.close()
+        done_file_obj.close()
+        err_file_obj.close()
+        sys.exit(1)
+
+
+class BlurredOriginalsDownloader(threading.Thread):
```

Comment:
Nice implementation of download threads! @jernejaMislej and I were thinking of using these for upload too